x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java

```diff
@@ -179,33 +207,69 @@ private void searchEventGroupByStackTrace(
-                // sort items lexicographically to access Lucene's term dictionary more efficiently when issuing an mget request.
+                StringTerms hosts = searchResponse.getAggregations().get("group_by");
+
+                // Create the list of host IDs and batch load the host metadata.
+                // TODO: use a global cache for the host metadata.
+                ArrayList<String> hostIDs = new ArrayList<>(hosts.getBuckets().size());
+                for (StringTerms.Bucket hostBucket : hosts.getBuckets()) {
+                    String hostid = hostBucket.getKeyAsString();
+                    hostIDs.add(hostid);
+                }
+                retrieveHostMetadata(client, hostIDs, responseBuilder, submitListener);
+
+                // Sort items lexicographically to access Lucene's term dictionary more efficiently when issuing an mget request.
                 // The term dictionary is lexicographically sorted and using the same order reduces the number of page faults
                 // needed to load it.
                 long totalFinalCount = 0;
-                Map<String, Integer> stackTraceEvents = new TreeMap<>();
-                for (StringTerms.Bucket bucket : stacktraces.getBuckets()) {
-                    Sum count = bucket.getAggregations().get("count");
-                    int finalCount = resampler.adjustSampleCount((int) count.value());
-                    totalFinalCount += finalCount;
-                    if (finalCount > 0) {
-                        stackTraceEvents.put(bucket.getKeyAsString(), finalCount);
+                Map<String, TraceEvent> stackTraceEvents = new TreeMap<>();
+                for (StringTerms.Bucket hostBucket : hosts.getBuckets()) {
+                    String hostid = hostBucket.getKeyAsString();
+                    StringTerms stacktraces = hostBucket.getAggregations().get("group_by");
+                    for (StringTerms.Bucket stacktraceBucket : stacktraces.getBuckets()) {
+                        Sum count = stacktraceBucket.getAggregations().get("count");
+                        log.debug("count value is " + count.value());
+                        int finalCount = resampler.adjustSampleCount((int) count.value());
+                        if (finalCount <= 0) {
+                            continue;
+                        }
+                        log.debug("final count value is " + finalCount);
+                        totalFinalCount += finalCount;
+                        /*
+                        The same stacktraces may come from different hosts (eventually from different datacenters).
+                        That is why we aggregate count, co2Factor and costFactor here.
+                         */
+                        String stackTraceID = stacktraceBucket.getKeyAsString();
+                        TraceEvent event = stackTraceEvents.get(stackTraceID);
+                        if (event == null) {
```

Comment:
TBH, I don't know how good the JIT compiler aggregates and optimizes this code... but the idiomatic code requires 2-3 tree walks, while the non-idiomatic code only requires 1-2.
Since we are in a relatively hot code path, I'd keep it as is, unless you have a reason not to do so.

Reply:
I'd favor readability over performance here. If it turns out to be a bottleneck we can always improve?

Reply:
TBH, how is `event == null` less readable than `stackTraceEvents.containsKey(stackTraceID) == false`?

Anyway, this is in a hot code path. I can't see the point in now making the code less readable, more complex, slower here.
Related: https://github.com/elastic/elasticsearch/pull/101979#discussion_r1388979638